src/vm/hints/usort.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{
+    bigint, bigintusize,
+    serde::deserialize_program::ApTracking,
+    types::exec_scope::PyValueType,
+    vm::{
+        errors::vm_errors::VirtualMachineError,
+        hints::hint_utils::{
+            get_int_from_scope, get_integer_from_relocatable_plus_offset,
+            get_integer_from_var_name, get_key_to_list_map_from_scope_mut, get_list_from_scope_mut,
+            get_list_from_scope_ref, get_range_check_builtin, get_relocatable_from_var_name,
+            insert_integer_at_relocatable_plus_offset, insert_integer_from_var_name,
+            insert_relocatable_from_var_name,
+        },
+        vm_core::VirtualMachine,
+    },
+};
+use num_bigint::BigInt;
+use num_traits::{FromPrimitive, ToPrimitive};
+use std::collections::HashMap;
+
+pub fn usort_enter_scope(vm: &mut VirtualMachine) -> Result<(), VirtualMachineError> {
+    let usort_max_size =
+        get_int_from_scope(vm, "usort_max_size").map_or(PyValueType::None, PyValueType::BigInt);
+    vm.exec_scopes.enter_scope(HashMap::from([(
+        "usort_max_size".to_string(),
+        usort_max_size,
+    )]));
+    Ok(())
+}
+
+pub fn usort_body(
+    vm: &mut VirtualMachine,
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let _ = get_range_check_builtin(vm)?;
+    let input_arr_start_ptr = get_relocatable_from_var_name("input", ids, vm, hint_ap_tracking)?;
+    let input_ptr = vm.memory.get_relocatable(&input_arr_start_ptr)?.clone();
+    let usort_max_size = get_int_from_scope(vm, "usort_max_size");
+    let input_len = get_integer_from_var_name("input_len", ids, vm, hint_ap_tracking)?;
+    let input_len_usize = input_len
+        .to_usize()
+        .ok_or(VirtualMachineError::BigintToUsizeFail)?;
+
+    if let Some(usort_max_size) = usort_max_size {
+        if input_len > &usort_max_size {
+            return Err(VirtualMachineError::UsortOutOfRange(
+                usort_max_size,
+                input_len.clone(),
+            ));
+        }
+    }
+
+    let mut positions_dict: HashMap<BigInt, Vec<BigInt>> = HashMap::new();
+    let mut output: Vec<BigInt> = Vec::new();
+    for i in 0..input_len_usize {
+        let val = get_integer_from_relocatable_plus_offset(&input_ptr, i, vm)?;
+        if let Err(output_index) = output.binary_search(val) {
+            output.insert(output_index, val.clone());
+        }
+        positions_dict
+            .entry(val.clone())
+            .or_insert(Vec::new())
+            .push(bigintusize!(i));
+    }
+
+    let mut multiplicities: Vec<BigInt> = Vec::new();
```

Comment:
We could use `u64` here as well, and only convert them when inserting it into memory.

Reply:
resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/319/commits/e560c5ca29b7dac63962c720a1a242e3c73fa040